Doc/library/pathlib.rst

```diff
@@ -680,6 +680,30 @@ Pure paths provide the following methods and properties:
       PureWindowsPath('README')
 
 
+.. method:: PurePath.__newpath__(*pathsegments)
+
+   Create a new path object of the same type by combining the given
+   *pathsegments*. This method is called whenever a derivative path is created,
+   such as from :attr:`parent` and :meth:`relative_to`. Subclasses may
+   override this method to pass information to derivative paths, for example::
+
+      from pathlib import PurePosixPath
+
+      class MyPath(PurePosixPath):
+          def __init__(self, *args, session_id):
+              super().__init__(*args)
+              self.session_id = session_id
+
+          def __newpath__(self, *pathsegments):
```

Comment:
The method should not use double underscores if it’s not a magic method (looked up on the class, `__init__` and `__class_getitem__` being the two exceptions)

Reply:
> `__init__` and `__class_getitem__` being the two exceptions

And `__mro_entries__`.

Is it documented anywhere that dunders _have_ to be looked up on the class? I know that many, if not most, are, but I thought that was "just" an optimisation.

Reply:
Far more than an optimization, It is part of the data model!
The opcode for `__enter__` did not do a proper lookup at first and was fixed in the next feature release.

https://docs.python.org/3/reference/datamodel.html#special-method-names

Reply:
I've adjusted the patch so `__newpath__()` is looked up on the class, not the instance.

Reply:
This method isn't a reserved name in Python's data model, so I wouldn't make it a dunder name. How about just `_newpath_()`? This is how `ctypes` names the internal attributes and methods of `ctypes` data objects, such as the `_check_retval_()` method. Plus then you don't have to worry about looking up the method on the type instead of just using `self._newpath_`.

Reply:
`enum` has a similar naming convention, using "sunders" for attributes that do something "special" inside the module, but aren't "reserved" names and aren't private either

Reply:
That's perfect, thanks :)

Reply:
Thought about this again – enum uses double single underscores, namedtuple single leading underscore because they need to allow arbitrary names for user-defined attributes.  There isn’t the same concert for Path subclasses, it would be fine to reserve one name `newpath` for this role, without conflict with `session_id` or `custom_method` needed by subclasses.

Reply:
Ack, right.

Having some sort of underscore prefix seems like a useful hint to users that they usually don't need to call this method, without precluding them from overriding it in a subclass. Without it, the method name needs to somehow hint at this. I don't like the way `path.newpath('blah')` reads, because it's not clear what role `path` plays, and several folks have asked "couldn't it be a classmethod?", which shows it's somewhat misleading.

I quite like how `path.with_segments('blah')` reads. It's less likely to be called by a new user who googles "pathlib new path". The "with" conjunction indicates that the path object may be referred to in some respect, and it's similar to the existing `with_name()` etc instance methods, so it seems less likely to be mistaken for a classmethod.